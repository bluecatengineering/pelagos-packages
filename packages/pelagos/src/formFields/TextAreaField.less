@import '../../less/fonts';
@import '../../less/focus';
@import '../../less/spacing';

.TextAreaField {
	margin-top: @sp-12;
}

.TextAreaField__area {
	height: 65px;
}

.TextAreaField--error {
	outline: 2px solid var(--support-error);
	outline-offset: -2px;
}

.TextAreaField--error:focus {
	@focus-ring-error();
}

.TextAreaField--noresize {
	resize: none;
}
