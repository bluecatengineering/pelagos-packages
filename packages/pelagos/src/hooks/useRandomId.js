import {useMemo} from 'react';

/**
 * Returns either the specified string or a stable random string.
 * @param {string} [id] the ID to use.
 * @returns {string} the ID.
 */
const useRandomId = (id) =>
	useMemo(() => id || `e${performance.now().toString(36)}${Math.random().toString(36).slice(1)}`, [id]);

export default useRandomId;
