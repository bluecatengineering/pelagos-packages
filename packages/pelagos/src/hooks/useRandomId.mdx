{/* DO NOT EDIT, generated from useRandomId.js */}

import {Meta} from '@storybook/addon-docs/blocks';

<Meta title="hooks/useRandomId" />

# useRandomId

Returns either the specified string or a stable random string.

## Parameters

<table>
	<thead>
		<tr>
			<th>Param</th>
			<th>Type</th>
			<th>Description</th>
		</tr>
	</thead>
	<tbody>
		<tr>
			<td>id</td>
			<td>`string`</td>
			<td>the ID to use.</td>
		</tr>
	</tbody>
</table>

## Returns

`string` - the ID.
