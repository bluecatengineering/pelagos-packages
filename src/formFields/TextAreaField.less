@import '../../less/colors';
@import '../../less/fonts';
@import '../../less/spacing';

.TextAreaField {
	margin-top: @sp-12;
}

.TextAreaField__label {
	flex-direction: row;
	margin-bottom: @sp-04;
}

.TextAreaField__optional {
	@font-10-400();
	color: @font-color-3;
	position: relative;
	top: -1px;
	margin-left: @sp-12;
}

.TextAreaField__area {
	height: 65px;
	padding: @sp-04 @sp-12;
	background-color: @form-field-bg;
	border: 1px solid @form-field-border;
	border-radius: 3px;
}

.TextAreaField--error {
	border-color: @form-field-error;
}

.TextAreaField--error:focus {
	outline-color: @form-field-error;
}

.TextAreaField--noresize {
	resize: none;
}

.Dialog {
	.TextAreaField__area {
		background-color: @oxford;
	}
}

@media print {
	.TextAreaField__area {
		background-color: transparent;
		border-color: transparent;
	}
}
