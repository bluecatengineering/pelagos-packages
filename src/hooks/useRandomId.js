import {useMemo} from 'react';

export default (id) => useMemo(() => id || `e${('' + Math.random()).slice(2)}`, [id]);
