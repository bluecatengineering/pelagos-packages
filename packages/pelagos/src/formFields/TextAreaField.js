import {useCallback, useMemo} from 'react';
import PropTypes from 'prop-types';
import debounce from 'lodash-es/debounce';

import Layer from '../components/Layer';
import LabelLine from '../components/LabelLine';
import useRandomId from '../hooks/useRandomId';

import FieldError from './FieldError';
import FieldHelper from './FieldHelper';
import './TextAreaField.less';

/** A text area field. */
const TextAreaField = ({id, className, label, value, optional, resize, helperText, error, onChange, ...props}) => {
	id = useRandomId(id);
	const helperId = `${id}-helper`;
	const errorId = `${id}-error`;
	const debounced = useMemo(() => debounce(onChange, 33), [onChange]);
	return (
		<Layer className={'TextAreaField' + (className ? ' ' + className : '')}>
			<LabelLine htmlFor={id} text={label} optional={optional && !value} />
			<textarea
				{...props}
				id={id}
				className={
					'TextAreaField__area' + (resize ? '' : ' TextAreaField--noresize') + (error ? ' TextAreaField--error' : '')
				}
				value={value}
				aria-describedby={error ? errorId : helperId}
				onChange={useCallback((event) => debounced(event.target.value), [debounced])}
			/>
			{error ? <FieldError id={errorId} text={error} /> : <FieldHelper id={helperId} text={helperText} />}
		</Layer>
	);
};

TextAreaField.propTypes = {
	/** The component ID. */
	id: PropTypes.string,
	/** The component class name(s). */
	className: PropTypes.string,
	/** The label text. */
	label: PropTypes.string.isRequired,
	/** The field value. */
	value: PropTypes.string,
	/** The placeholder text. */
	placeholder: PropTypes.string,
	/** Whether the field is optional. */
	optional: PropTypes.bool,
	/** Whether the field can be resized. */
	resize: PropTypes.bool,
	/** The maximum text length. */
	maxLength: PropTypes.number,
	/** Additional information for the field. */
	helperText: PropTypes.string,
	/** The error text. */
	error: PropTypes.string,
	/** Function invoked when the value changes. */
	onChange: PropTypes.func.isRequired,
};

export default TextAreaField;
