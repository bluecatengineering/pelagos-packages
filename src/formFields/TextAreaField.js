import React from 'react';
import PropTypes from 'prop-types';

import Label from '../components/Label';

import FieldError from './FieldError';
import './TextAreaField.less';

const TextAreaField = ({id, className, label, value, optionalText, resize, error, ...props}) => {
	id = id || 'e' + ('' + Math.random()).substr(2);
	return (
		<div className={'TextAreaField' + (className ? ' ' + className : '')}>
			<div className="TextAreaField__label">
				<Label text={label} htmlFor={id} />
				{optionalText && !value && <span className="TextAreaField__optional">{optionalText}</span>}
			</div>
			<textarea
				{...props}
				id={id}
				className={
					'TextAreaField__area' + (resize ? '' : ' TextAreaField--noresize') + (error ? ' TextAreaField--error' : '')
				}
				value={value}
			/>
			<FieldError text={error} />
		</div>
	);
};

TextAreaField.propTypes = {
	id: PropTypes.string,
	className: PropTypes.string,
	label: PropTypes.string.isRequired,
	value: PropTypes.string,
	placeholder: PropTypes.string,
	optionalText: PropTypes.string,
	resize: PropTypes.bool,
	maxLength: PropTypes.number,
	error: PropTypes.string,
	onChange: PropTypes.func.isRequired,
};

export default TextAreaField;
