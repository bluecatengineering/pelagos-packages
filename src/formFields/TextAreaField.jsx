import React from 'react';
import PropTypes from 'prop-types';

import Label from '../components/Label';

import FieldError from './FieldError';
import './TextAreaField.less';

const TextAreaField = ({id, className, label, resize, error, ...props}) => {
	id = id || 'e' + ('' + Math.random()).substr(2);
	return (
		<div className={'TextAreaField' + (className ? ' ' + className : '')}>
			<div>
				<Label text={label} htmlFor={id} />
			</div>
			<textarea
				{...props}
				id={id}
				className={
					'TextAreaField__area' + (resize ? '' : ' TextAreaField--noresize') + (error ? ' TextAreaField--error' : '')
				}
			/>
			<FieldError text={error} />
		</div>
	);
};

TextAreaField.propTypes = {
	id: PropTypes.string,
	className: PropTypes.string,
	label: PropTypes.string.isRequired,
	value: PropTypes.string,
	placeholder: PropTypes.string,
	resize: PropTypes.bool,
	maxLength: PropTypes.number,
	error: PropTypes.string,
	onChange: PropTypes.func.isRequired,
};

export default TextAreaField;
