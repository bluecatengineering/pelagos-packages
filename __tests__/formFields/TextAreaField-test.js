import {shallow} from 'enzyme';

import TextAreaField from '../../src/formFields/TextAreaField';
import useRandomId from '../../src/hooks/useRandomId';

jest.unmock('../../src/formFields/TextAreaField');
jest.mock('lodash-es/debounce', () => jest.fn((f) => f));

useRandomId.mockReturnValue('test-id');

describe('TextAreaField', () => {
	describe('rendering', () => {
		it('renders expected elements', () => {
			const wrapper = shallow(
				<TextAreaField
					id="test-id"
					label="Label"
					value="value"
					placeholder="placeholder"
					optional
					maxLength={10}
					onChange={jest.fn()}
				/>
			);
			expect(wrapper.getElement()).toMatchSnapshot();
		});

		it('renders expected elements when value is empty', () => {
			const wrapper = shallow(
				<TextAreaField
					id="test-id"
					label="Label"
					value=""
					placeholder="placeholder"
					optional
					maxLength={10}
					onChange={jest.fn()}
				/>
			);
			expect(wrapper.getElement()).toMatchSnapshot();
		});

		it('renders expected elements when className is set', () => {
			const wrapper = shallow(
				<TextAreaField
					id="test-id"
					className="TestClass"
					label="Label"
					value="value"
					placeholder="placeholder"
					maxLength={10}
					onChange={jest.fn()}
				/>
			);
			expect(wrapper.getElement()).toMatchSnapshot();
		});

		it('does not add the resize class if the resize is true', () => {
			const wrapper = shallow(
				<TextAreaField
					id="test-id"
					label="Label"
					value="value"
					placeholder="placeholder"
					maxLength={10}
					resize={true}
					onChange={jest.fn()}
				/>
			);
			expect(wrapper.getElement()).toMatchSnapshot();
		});

		it('adds the error class if the error is set', () => {
			const wrapper = shallow(
				<TextAreaField
					id="test-id"
					label="Label"
					value="value"
					placeholder="placeholder"
					maxLength={10}
					error="Error"
					onChange={jest.fn()}
				/>
			);
			expect(wrapper.getElement()).toMatchSnapshot();
		});
	});

	describe('behaviour', () => {
		it('calls onChange when a change event is fired', () => {
			const onChange = jest.fn();
			const event = {target: {value: 'test'}};
			const wrapper = shallow(<TextAreaField label="Label" value="" onChange={onChange} />);
			wrapper.find('textarea').simulate('change', event);
			expect(onChange.mock.calls).toEqual([['test']]);
		});
	});
});
