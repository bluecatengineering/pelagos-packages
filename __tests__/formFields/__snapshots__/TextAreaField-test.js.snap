// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextAreaField rendering adds the error class if the error is set 1`] = `
<div
  className="TextAreaField"
>
  <LabelLine
    htmlFor="test-id"
    text="Label"
  />
  <textarea
    className="TextAreaField__area TextAreaField--noresize TextAreaField--error"
    id="test-id"
    maxLength={10}
    onChange={[Function]}
    placeholder="placeholder"
    value="value"
  />
  <FieldError
    alignment="right"
    text="Error"
  />
</div>
`;

exports[`TextAreaField rendering does not add the resize class if the resize is true 1`] = `
<div
  className="TextAreaField"
>
  <LabelLine
    htmlFor="test-id"
    text="Label"
  />
  <textarea
    className="TextAreaField__area"
    id="test-id"
    maxLength={10}
    onChange={[Function]}
    placeholder="placeholder"
    value="value"
  />
  <FieldError
    alignment="right"
  />
</div>
`;

exports[`TextAreaField rendering renders expected elements 1`] = `
<div
  className="TextAreaField"
>
  <LabelLine
    htmlFor="test-id"
    optional={false}
    text="Label"
  />
  <textarea
    className="TextAreaField__area TextAreaField--noresize"
    id="test-id"
    maxLength={10}
    onChange={[Function]}
    placeholder="placeholder"
    value="value"
  />
  <FieldError
    alignment="right"
  />
</div>
`;

exports[`TextAreaField rendering renders expected elements when className is set 1`] = `
<div
  className="TextAreaField TestClass"
>
  <LabelLine
    htmlFor="test-id"
    text="Label"
  />
  <textarea
    className="TextAreaField__area TextAreaField--noresize"
    id="test-id"
    maxLength={10}
    onChange={[Function]}
    placeholder="placeholder"
    value="value"
  />
  <FieldError
    alignment="right"
  />
</div>
`;

exports[`TextAreaField rendering renders expected elements when value is empty 1`] = `
<div
  className="TextAreaField"
>
  <LabelLine
    htmlFor="test-id"
    optional={true}
    text="Label"
  />
  <textarea
    className="TextAreaField__area TextAreaField--noresize"
    id="test-id"
    maxLength={10}
    onChange={[Function]}
    placeholder="placeholder"
    value=""
  />
  <FieldError
    alignment="right"
  />
</div>
`;

exports[`TextAreaField rendering sets a random id if not provided 1`] = `
<div
  className="TextAreaField"
>
  <LabelLine
    htmlFor="e1"
    text="Label"
  />
  <textarea
    className="TextAreaField__area TextAreaField--noresize"
    id="e1"
    maxLength={10}
    onChange={[Function]}
    placeholder="placeholder"
    value="value"
  />
  <FieldError
    alignment="right"
  />
</div>
`;
